File: Java/Linked List/Singly Circular Linked List/SinglyCL19(FINAL).java
/*
=====================================================================
 
Name            : Singly Circular Linked List - Complete Implementation
Program Number  : 019

Description     : This program implements a complete Singly Circular
                  Linked List using Java.

                  The application provides a menu-driven interface
                  to perform various operations on the linked list.

                  Supported Operations:

                  - InsertFirst()
                  - InsertLast()
                  - InsertAtPos()
                  - DeleteFirst()
                  - DeleteLast()
                  - DeleteAtPos()
                  - Display()
                  - Count()

                  The circular nature of the linked list is maintained
                  using first and last references where last.next
                  always points to first.

Concepts Used   : Singly Circular Linked List,
                  Dynamic Memory Allocation,
                  Object Oriented Programming,
                  Constructor,
                  Traversal,
                  Insertion,
                  Deletion,
                  Position Based Operations,
                  Menu Driven Programming

Input           : User choice, node data and position

Output          : Displays linked list and operation results

Date            : July 2026
Language        : Java
Author          : Atharv Tushar Bhosale

=====================================================================
*/

import java.util.*;

class node
{
    public int data;
    public node next;

    node(int no)
    {
        this.data = no;
        this.next = null;
    }
}

class SinglyCL
{
    public node first;
    public node last;
    public int iCount;

    public SinglyCL()
    {
        this.first = null;
        this.last = null;
        this.iCount = 0;
    }

    public void Display()
    {
        node temp = null;
        temp = first;

        if(first == null && last == null)
        {
            return;
        }

        System.out.print("->");
        do
        {
            System.out.print("| "+temp.data+" | -> ");
            temp = temp.next;
        }while(temp != last.next);
        System.out.println();
    }

    public int Count()
    {
        return  iCount;
    }

    public void InsertFirst(int no)
    {
        node newn= new node(no);

        if(first == null && last == null)
        {
            first = newn;
            last = newn;
        }
        else
        {
            newn.next = first;
            first = newn;
        }
        last.next = first;
        iCount++;
    }

    public void InsertLast(int no)
    {
        node newn= new node(no);

        if(first == null && last == null)
        {
            first = newn;
            last = newn;
        }
        else
        {
            last.next = newn;
            last = newn;
        }
        last.next = first;
        iCount++;
    }

    public void InsertAtPos(int no, int iPos)
    {
        int i = 0;
        node temp = null;
        node newn = new node(no);

        if(iPos < 1 || iPos > iCount + 1)
        {
            return;
        }

        if(iPos == 1)
        {
            InsertFirst(no);
        }
        else if(iPos == iCount + 1)
        {
            InsertLast(no);
        }
        else
        {
            temp = first;

            for(i = 1; i < iPos - 1; i++)
            {
                temp = temp.next;
            }
            newn.next = temp.next;
            temp.next = newn;
            iCount++;
        }
        
    }

    public void DeleteFirst()
    {
        if(first == null && last == null)
        {
            return;
        }
        else if(first == last)
        {
            first = null;
            last = null;
        }
        else
        {
            first = first.next;
            last.next = first;
        }
        iCount--;
    }

    public void DeleteLast()
    {
        if(first == null && last == null)
        {
            return;
        }
        else if(first == last)
        {
            first = null;
            last = null;
        }
        else
        {
            node temp = null;

            temp = first;

            while(temp.next != last)
            {
                temp = temp.next;
            }
            last = temp;
            last.next = first;
        }
        iCount--;
    }

    public void DeleteAtPos(int iPos)
    {
        int i = 0;
        node temp = null;
        node target = null;

        if(iPos < 1 || iPos > iCount)
        {
            return;
        }

        if(iPos == 1)
        {
            DeleteFirst();
        }
        else if(iPos == iCount)
        {
            DeleteLast();
        }
        else
        {
            temp = first;

            for(i = 1; i < iPos - 1; i++)
            {
                temp = temp.next;
            }
            target = temp.next;
            temp.next = target.next;
            iCount--;
        }
    }
}



class SinglyCL17
{
    public static void main(String A[])
    {
        Scanner sobjScanner = new Scanner(System.in);

        SinglyCL sobj = new SinglyCL();

        int iChoice = 0;
        int iNo = 0;
        int iPos = 0;

        while(true)
        {
            System.out.println("\n-----------------------------");
            System.out.println("Singly Circular Linked List");
            System.out.println("-----------------------------");
            System.out.println("1 : Insert First");
            System.out.println("2 : Insert Last");
            System.out.println("3 : Insert At Position");
            System.out.println("4 : Delete First");
            System.out.println("5 : Delete Last");
            System.out.println("6 : Delete At Position");
            System.out.println("7 : Display");
            System.out.println("8 : Count");
            System.out.println("0 : Exit");
            System.out.print("Enter your choice : ");
            iChoice = sobjScanner.nextInt();

            switch(iChoice)
            {
                case 1:
                    System.out.print("Enter data : ");
                    iNo = sobjScanner.nextInt();
                    sobj.InsertFirst(iNo);
                    break;

                case 2:
                    System.out.print("Enter data : ");
                    iNo = sobjScanner.nextInt();
                    sobj.InsertLast(iNo);
                    break;

                case 3:
                    System.out.print("Enter data : ");
                    iNo = sobjScanner.nextInt();

                    System.out.print("Enter position : ");
                    iPos = sobjScanner.nextInt();

                    sobj.InsertAtPos(iNo, iPos);
                    break;

                case 4:
                    sobj.DeleteFirst();
                    break;

                case 5:
                    sobj.DeleteLast();
                    break;

                case 6:
                    System.out.print("Enter position : ");
                    iPos = sobjScanner.nextInt();

                    sobj.DeleteAtPos(iPos);
                    break;

                case 7:
                    sobj.Display();
                    break;

                case 8:
                    System.out.println("Number of Nodes : " + sobj.Count());
                    break;

                case 0:
                    System.out.println("Thank you for using the application");
                    sobjScanner.close();
                    return;

                default:
                    System.out.println("Invalid choice");
                    break;
            }
        }
    }
}

/*
=====================================================================

Output :
Successfully performs basic operations on Singly Circular Linked List.

Operations Available :
- Insert First
- Insert Last
- Insert At Position
- Delete First
- Delete Last
- Delete At Position
- Display Nodes
- Count Nodes

Concepts Covered :
- Circular Linked List
- Dynamic Memory Allocation using Objects
- Node Traversal using do-while Loop
- Menu Driven Programming
- Object Oriented Programming in Java

Result :
All basic Singly Circular Linked List operations executed successfully.

=====================================================================
*/

File: C++/Linked List/Singly Circular Linked List/SinglyCL19(FINAL).cpp
/*
=====================================================================

Name            : Singly Circular Linked List - Menu Driven Program
Program Number  : 022

Description     : This program demonstrates a menu driven
                  implementation of Singly Circular Linked List.

                  User can perform the following operations:

                  1. InsertFirst()
                  2. InsertLast()
                  3. InsertAtPos()
                  4. DeleteFirst()
                  5. DeleteLast()
                  6. DeleteAtPos()
                  7. Display()
                  8. Count()

Concepts Used   : Classes and Objects,
                  Dynamic Memory Allocation,
                  Circular Linked List,
                  Menu Driven Programming

Input           : User Choice and Node Data

Output          : Updated linked list after every operation

Date            : July 2026
Language        : C++
Author          : Atharv Tushar Bhosale

=====================================================================
*/

#include<iostream>
using namespace std;

#pragma pack(1)

struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node* PNODE;

class SinglyCL
{
    private:
        PNODE first;
        PNODE last;
        int iCount;
        
    public:
        SinglyCL();
        void Display();
        int Count();

        void InsertFirst(int iNo);
        void InsertLast(int iNo);
        void InsertAtPos(int iNo, int iPos);

        void DeleteFirst();
        void DeleteLast();
        void DeleteAtPos(int iPos);
};

SinglyCL :: SinglyCL()
{
    cout<<"Inside Constructor\n";
    this -> first = NULL;
    this -> last = NULL;
    this -> iCount = 0;
}

void SinglyCL :: Display()
{
    PNODE temp = NULL;

    if(first == NULL && last == NULL)
    {
        return;
    }

    temp = first;

    do
    {
        cout<<"| "<<temp -> data<<" | ->";
        temp = temp -> next;
    }while(last -> next != temp);
    cout<<"\n";

}

int SinglyCL :: Count()
{
    return iCount;
}

void SinglyCL :: InsertFirst(int iNo)
{
    PNODE newn = NULL;

    newn = new NODE();

    newn -> data = iNo;
    newn -> next = NULL;

    if(first == NULL && last == NULL)
    {
        first = newn;
        last = newn;
    }
    else
    {
        newn -> next = first;
        first = newn;
        
    }
    last -> next = first;
    iCount++;
}

void SinglyCL :: InsertLast(int iNo)
{
    PNODE newn = NULL;

    newn = new NODE();

    newn -> data = iNo;
    newn -> next = NULL;

    if(first == NULL && last == NULL)
    {
        first = newn;
        last = newn;
    }
    else
    {
        last -> next = newn;
        last = newn;
    }
    last -> next = first;
    iCount++;
}

void SinglyCL ::InsertAtPos(int iNo, int iPos)
{
    int i = 0;

    if(iPos < 1 || iPos > iCount + 1)
    {
        cout<<"Invalid Positions\n";
        return;
    }

    if(iPos == 1)
    {
        InsertFirst(iNo);
    }
    else if(iPos == iCount + 1)
    {
        InsertLast(iNo);
    }
    else
    {
        PNODE temp = NULL;
        PNODE newn = NULL;

        newn = new NODE();

        newn -> data = iNo;
        newn -> next = NULL;

        temp = first;
        for(i = 1; i < iPos - 1; i++)
        {
            temp = temp -> next;
        }

        newn -> next = temp -> next;
        temp -> next = newn;
        iCount++;
    }
}

void SinglyCL :: DeleteFirst()
{
    if(first == NULL && last == NULL)
    {
        return;
    }
    else if(first == last)
    {
        delete first;
        first = NULL;
        last = NULL;
    }
    else
    {
        first = first -> next;
        delete last -> next;
        last -> next = first;
    }
    iCount--;
}

void SinglyCL :: DeleteLast()
{
    if(first == NULL && last == NULL)
    {
        return;
    }
    else if(first == last)
    {
        delete first;
        first = NULL;
        last = NULL;
    }
    else
    {
        PNODE temp = NULL;

        temp = first;

        while(temp -> next != last)
        {
            temp = temp -> next;
        }
        delete last;
        last = temp;
        last -> next = first;
    }
    iCount--;
}

void SinglyCL :: DeleteAtPos(int iPos)
{
    int i = 0;

    if(iPos < 1 || iPos > iCount)
    {
        cout<<"Invalid Positions\n";
        return;
    }

    if(iPos == 1)
    {
        DeleteFirst();
    }
    else if(iPos == iCount)
    {
        DeleteLast();
    }
    else
    {
        PNODE temp = NULL;
        PNODE target = NULL;

        temp = first;

        for(i = 1; i < iPos - 1; i++)
        {
            temp = temp -> next;
        }

        target = temp -> next;
        temp -> next = target -> next;
        delete target;
        iCount--;
    }
}

int main()
{
    SinglyCL sobj;

    int iChoice = 0;
    int iValue = 0;
    int iPos = 0;
    int iRet = 0;

    while(true)
    {
        cout<<"\n-----------------------------------\n";
        cout<<"    Singly Circular Linked List\n";
        cout<<"-----------------------------------\n";

        cout<<"1 : Insert First\n";
        cout<<"2 : Insert Last\n";
        cout<<"3 : Insert At Position\n";
        cout<<"4 : Delete First\n";
        cout<<"5 : Delete Last\n";
        cout<<"6 : Delete At Position\n";
        cout<<"7 : Display\n";
        cout<<"8 : Count\n";
        cout<<"0 : Exit\n";

        cout<<"Enter your choice : ";
        cin>>iChoice;

        switch(iChoice)
        {
            case 1:
                cout<<"Enter data : ";
                cin>>iValue;
                sobj.InsertFirst(iValue);
                break;

            case 2:
                cout<<"Enter data : ";
                cin>>iValue;
                sobj.InsertLast(iValue);
                break;

            case 3:
                cout<<"Enter data : ";
                cin>>iValue;

                cout<<"Enter position : ";
                cin>>iPos;

                sobj.InsertAtPos(iValue,iPos);
                break;

            case 4:
                sobj.DeleteFirst();
                break;

            case 5:
                sobj.DeleteLast();
                break;

            case 6:
                cout<<"Enter position : ";
                cin>>iPos;

                sobj.DeleteAtPos(iPos);
                break;

            case 7:
                sobj.Display();
                break;

            case 8:
                iRet = sobj.Count();
                cout<<"Number of Nodes are : "<<iRet<<"\n";
                break;

            case 0:
                cout<<"Thank you for using the application\n";
                return 0;

            default:
                cout<<"Invalid choice\n";
                break;
        }
    }

    return 0;
}

/*
=====================================================================

Program Summary :

-> Implemented Singly Circular Linked List
-> Performed insertion operations
-> Performed deletion operations
-> Displayed linked list elements
-> Counted total number of nodes
-> Used menu driven approach for user interaction

=====================================================================
*/
